### PR TITLE
Add file list filter to filter by system tags

### PR DESCRIPTION
**Pull request by PVince81**
- [x] REQUIRES: backend REPORT method for files https://github.com/owncloud/core/pull/22112
- [x] add JS unit tests
- [ ] add icons
- [ ] fix ugly styling
- [x] TODO: allow other sidebar plugins to register to this list (system tags)
- [x] test all file actions to make sure they work properly in this view
- [x] test sidebar actions to make sure they work properly in this view

Ref: https://github.com/owncloud/core/issues/21001

Note: the backend has some bugs so some properties are wrong (size), see https://github.com/owncloud/core/pull/22112#issuecomment-179482473

systemtags-filter-gui

### CODE REVIEW

**Comment by PVince81** [comment]
- [ ] REQUIRES https://github.com/owncloud/core/pull/22213 for the share dialog to work properly when the file is in a subdir

**Comment by karlitschek** [comment]
Awesome @PVince81 :+1:

**Comment by PVince81** [comment]
- [ ] TODO: add tag ids in the URL (since we don't have presets, at least being able to bookmark is nice)

**Comment by PVince81** [comment]
Now (ab)using the "dir" URL attribute to store tag ids. This makes it possible to bookmark the page and return to the filtered list later.

**Comment by PVince81** [comment]
- [ ] BUG: text editor broken on this branch ?!!

**Comment by PVince81** [comment]
@jancborchardt can you review this UX-wise and provide an icon ? (separately is fine too)

**Comment by PVince81** [comment]
For now I'll copy the search icon and call it "tag.svg" for the sake of getting this PR merged quickly.
It can later be replaced with a better icon, maybe one that looks like this: http://fortawesome.github.io/Font-Awesome/icon/tag/

**Comment by PVince81** [comment]
Now it looks like this:
systemtags-filter-gui2

If you want to change the icon now, it's in "apps/systemtags/img/tag.{svg,png}"

Please review @jancborchardt @rullzer @nickvergessen @MorrisJobke @icewind1991 @blizzz

File: apps/systemtags/appinfo/app.php [hunk 1]
```diff
@@ -73,3 +75,15 @@ function() {
 
 $eventDispatcher->addListener(MapperEvent::EVENT_ASSIGN, $mapperListener);
 $eventDispatcher->addListener(MapperEvent::EVENT_UNASSIGN, $mapperListener);
+
+$l = \OC::$server->getL10N('files_sharing');
```
Comment (rullzer) [inline apps/systemtags/appinfo/app.php]: That does not look right...

**Comment by DeepDiver1975** [comment]
@PVince81 

````
05:04:25 There was 1 failure:
05:04:25 
05:04:25 1) OCA\Files\Tests\Controller\ViewControllerTest::testIndexWithRegularBrowser
05:04:25 Failed asserting that two objects are equal.
05:04:25 --- Expected
05:04:25 +++ Actual
05:04:25 @@ @@
05:04:25                      2 => Array (
05:04:25 +                        'id' => 'systemtagsfilter'
05:04:25 +                        'appname' => 'systemtags'
05:04:25 +                        'script' => 'list.php'
05:04:25 +                        'order' => 9
05:04:25 +                        'name' => 'Tags'
05:04:25 +                        'active' => false
05:04:25 +                        'icon' => ''
05:04:25 +                    )
05:04:25 +                    3 => Array (
05:04:25                          'id' => 'sharingin'
05:04:25                          'appname' => 'files_sharing'
05:04:25                          'script' => 'list.php'
05:04:25                          'order' => 10
05:04:25 -                        'name' => OC_L10N_String Object (...)
05:04:25 +                        'name' => 'Shared with you'
05:04:25                          'active' => false
05:04:25                          'icon' => ''
05:04:25                      )
05:04:25 -                    3 => Array (
05:04:25 +                    4 => Array (
05:04:25                          'id' => 'sharingout'
05:04:25                          'appname' => 'files_sharing'
05:04:25                          'script' => 'list.php'
05:04:25                          'order' => 15
05:04:25 -                        'name' => OC_L10N_String Object (...)
05:04:25 +                        'name' => 'Shared with others'
05:04:25                          'active' => false
05:04:25                          'icon' => ''
05:04:25                      )
05:04:25 -                    4 => Array (
05:04:25 +                    5 => Array (
05:04:25                          'id' => 'sharinglinks'
05:04:25                          'appname' => 'files_sharing'
05:04:25                          'script' => 'list.php'
05:04:25                          'order' => 20
05:04:25 -                        'name' => OC_L10N_String Object (...)
05:04:25 +                        'name' => 'Shared by link'
05:04:25                          'active' => false
05:04:25                          'icon' => ''
05:04:25                      )
05:04:25 -                    5 => Array (
05:04:25 -                        'id' => 'trashbin'
05:04:25 -                        'appname' => 'files_trashbin'
05:04:25 -                        'script' => 'list.php'
05:04:25 -                        'order' => 50
05:04:25 -                        'name' => OC_L10N_String Object (...)
05:04:25 -                        'active' => false
05:04:25 -                        'icon' => ''
05:04:25 -                    )
05:04:25 +                    6 => Array (...)
05:04:25 
05:04:25 @@ @@
05:04:25              2 => Array (
05:04:25 -                'id' => 'sharingin'
05:04:25 +                'id' => 'systemtagsfilter'
05:04:25                  'content' => null
05:04:25              )
05:04:25              3 => Array (
05:04:25 -                'id' => 'sharingout'
05:04:25 +                'id' => 'sharingin'
05:04:25                  'content' => null
05:04:25              )
05:04:25              4 => Array (
05:04:25 -                'id' => 'sharinglinks'
05:04:25 +                'id' => 'sharingout'
05:04:25                  'content' => null
05:04:25              )
05:04:25              5 => Array (
05:04:25 -                'id' => 'trashbin'
05:04:25 +                'id' => 'sharinglinks'
05:04:25                  'content' => null
05:04:25              )
05:04:25 +            6 => Array (...)
05:04:25          )
05:04:25      )
05:04:25      'renderAs' => 'user'
05:04:25      'appName' => 'files'
05:04:25      'headers' => Array (...)
05:04:25      'cookies' => Array ()
05:04:25      'status' => 200
05:04:25      'lastModified' => null
05:04:25      'ETag' => null
05:04:25      'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
05:04:25  )
05:04:25

**Comment by karlitschek** [comment]
i think it looks ok. or do you prefere a tag icon @jancborchardt  ?

**Comment by PVince81** [comment]
@karlitschek I'd prefer a tag icon but am not willing to get my hands dirty with inkscape or searching the web for compliant icons. Leaving this up to @jancborchardt 

@DeepDiver1975 I'll fix the unit tests, was so focussed on JS that I forgot that there was PHP too

**Comment by PVince81** [comment]
Fixed unit test + l10n copy-paste fail.

File: apps/files/tests/controller/ViewControllerTest.php [hunk 2]
```diff
@@ -168,6 +168,15 @@ public function testIndexWithRegularBrowser() {
 				'icon' => '',
 			],
 			2 => [
+			'id' => 'systemtagsfilter',
+				'appname' => 'systemtags',
+				'script' => 'list.php',
+				'order' => 9,
+				'name' => new \OC_L10N_String(new \OC_L10N('systemtags'), 'Tags', []),
```
Comment (nickvergessen) [inline apps/files/tests/controller/ViewControllerTest.php]: No, no, no :cry: 
Apps should not use private stuff, since 9.0 this is a string only.

**Comment by nickvergessen** [comment]
JS UI works :+1: 

the endpoint has an issue thou, as per RL-discussion

**Comment by PVince81** [comment]
@nickvergessen fix for backend bug is here https://github.com/owncloud/core/pull/22234

File: apps/files/tests/controller/ViewControllerTest.php (hunk 2 above)
Comment (PVince81) [inline apps/files/tests/controller/ViewControllerTest.php]: I just copy pasted, and this is a unit test. See https://github.com/owncloud/core/pull/22197/files#diff-929f460429d7eed2f7cfe3b2c2fcd8ecR157

**Comment by rullzer** [comment]
:+1:

**Comment by PVince81** [comment]
Once this is merged I'll make a separate ticket to fix the design / icons

File: apps/files/tests/controller/ViewControllerTest.php (hunk 2 above)
Comment (nickvergessen) [inline apps/files/tests/controller/ViewControllerTest.php]: Okay, I will fix this then, once your PR is in

**Comment by PVince81** [comment]
Raised https://github.com/owncloud/core/issues/22240 to have a look at improving the design

**Pull request merged**